### ISSUE TITLE
Doesn't work at all (anymore, used to)

### ISSUE THREAD

**Issue by ghost** [mps-youtube/mps-youtube#752]
Hello,

## Issue / Suggestion
- `rm -r ~/.config/mps-youtube/` to be sure
- Launch and try any video
- It doesn't work (no crash) and acts as if nothing happened (see video: https://a.pomf.space/vgoutmsjmdje.webm)
- Trying to use the command line specified in the log works fine.

log: https://a.pomf.space/dceoiuhvolip.log

## Your Environment

```
mpsyt version      : 0.2.7.1
   notes           : released 6 July 2016
pafy version       : 0.5.3.1 (youtube-dl backend)
youtube-dl version : 2017.12.31
Python version     : 3.5.4 (default, Dec  7 2017, 17:07:05)
[GCC 6.4.0]
Processor          : AMD FX(tm)-8350 Eight-Core Processor
Machine type       : x86_64
Architecture       : 64bit,
Platform           : Linux-4.14.8-gentoo-r1-x86_64-AMD_FX-tm-8350_Eight-Core_Processor-with-gentoo-2.4.1
sys.stdout.enc     : UTF-8
default enc        : utf-8
Config dir         : /home/user/.config/mps-youtube
env:TERM           : st-256color
env:SHELL          : /bin/zsh
env:LANG           : en_US.utf8
```
Gentoo GNU/Linux, using mpv 0.27, ffmpeg 3.4.1.

**Comment by ritiek** [comment]
Are you able to play media directly using `mpv`?

```
mpv https://www.youtube.com/watch?v=bM7SZ5SBzyY
```

**Comment by Q3CPMA** [comment]
Yes, as I said, the command line called in the log, `mpv --title Mussorgsky - Pictures at an Exhibition --no-ytdl https://r2---sn-25ge7nsk.googlevideo.com/videoplayback?gir=yes&mm=31&mn=sn-25ge7nsk&mt=1515010135&mv=m&ei=oDhNWoaXNY61WNf6mfAI&ms=au&requiressl=yes&key=yt6&ip=83.200.90.26&lmt=1449557499383412&dur=2013.201&expire=1515031809&source=youtube&initcwndbps=593750&id=o-AIZYKmLEu-ZZr5P6g7BthaHkjjcbazLXZNbgq5Tjw6d5&sparams=clen%2Cdur%2Cei%2Cgir%2Cid%2Cinitcwndbps%2Cip%2Cipbits%2Citag%2Ckeepalive%2Clmt%2Cmime%2Cmm%2Cmn%2Cms%2Cmv%2Cpl%2Crequiressl%2Csource%2Cexpire&ipbits=0&clen=34708221&signature=7B5B36DAFC00AA241E735156D53E1FD1EF154953.6E1DF136C53E8CCDA15401247FFD223B7879E01E&mime=audio%2Fwebm&pl=17&itag=251&keepalive=yes&ratebypass=yes` (with the proper quoting) works as it should.

**Comment by Razesdark** [comment]
Yeah. The it seems quoting is the issue, here. Running that same command with title in quotes seemed to atleast get the the player to try to play the link.

**Comment by uNouss** [comment]
I have the issue that I report here #744

**Comment by Q3CPMA** [comment]
Related to this issue (I think): there should be a way to see the output
(stdout+stderr) of the mpv command used for debugging purposes.

**Comment by jahlborn** [comment]
Having this same issue.  may songs just don't play.  using the url directly from the log file doesn't work because it isn't properly quoted.  when i quote it, it works fine.  playing the youtube url directly with mpv also works fine.  any progress on this?

**Comment by vn-ki** [comment]
Can you try the latest code in `develop`? This should be fixed by b50f788.

**Comment by Q3CPMA** [comment]
I tried using git (with pafy git too) today, and while it doesn't work, it
doesn't skip as in my first video. Now it loops/blink.

**Comment by vn-ki** [comment]
@Q3CPMA Can you provide a log? The old one is not there anymore.

**Comment by Q3CPMA** [comment]
On Mon, Feb 19, 2018 at 12:55:04PM -0800, Vishnunarayan K I wrote:
> @Q3CPMA Can you provide a log? The old one is not there anymore.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/mps-youtube/mps-youtube/issues/752#issuecomment-366800546

Sure, here is one. http://0x0.st/sTdQ.log

**Comment by vn-ki** [comment]
> (with the proper quoting) works as it should.

@Q3CPMA Can you describe how you quoted the command? I can't reproduce this issue on my machine. 

In my opinion, quoting shouldn't be a problem because `subprocess.Popen` accept a list. The actual executed code is `subprocess.Popen(['mpv', '--title', '<title>' ... ])`.

**Comment by jahlborn** [comment]
Same, i get the loop/blink issue.

**Comment by vn-ki** [comment]
@jahlborn Please specify which version you were using. The stable version (installed by `pip` or from `master`) or the `develop` version? The one in `develop` might have some unidentified bugs causing the loop issue (due to refactoring of player.py #753). 

Please provide a log in either case. (Open mpsyt with `mpsyt --debug`. The log can be found in `/tmp/mpsyt.log`)

**Comment by jahlborn** [comment]
sorry, i updated to the develop version as you suggested.

**Comment by vn-ki** [comment]
@jahlborn Do you have the problem with `develop` branch too? If yes, please provide a log.

**Comment by jahlborn** [comment]
i do.  i had a log but i already deleted it and reverted to the release version.  my log file looked the same as the one provided by @Q3CPMA

**Comment by ritiek** [comment]
Let's check if this problem is really about quotes. What happens when you run this python code:
```python
>>> import subprocess
>>> # url from log
>>> url = 'https://r2---sn-ci5gup-qxaz.googlevideo.com/videoplayback?fvip=2...'
>>> cmd = ['mpv', '--no-ytdl', url]
>>> subprocess.Popen(cmd)
```
Does this play the media correctly? If not, does it give any error?

**Comment by jahlborn** [comment]
that python code works fine for me.  plays the song without issue.

**Comment by Q3CPMA** [comment]
On Tue, Feb 20, 2018 at 10:53:17AM +0000, Vishnunarayan K I wrote:
> > (with the proper quoting) works as it should.
>
> @Q3CPMA Can you describe how you quoted the command? I can't reproduce this issue on my machine.
>
> In my opinion, quoting shouldn't be a problem because `subprocess.Popen` accept a list. The actual executed code is `subprocess.Popen(['mpv', '--title', '<title>' ... ])`.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/mps-youtube/mps-youtube/issues/752#issuecomment-366940350

As you can see in the new log, what caused issues (the title) is now quoted.
Pasting the cmd in any shell (while quoting the URL, because of all the &s)
works perfectly.

This is the work command line, with single quotes added, mpv --title "Mussorgsky - Pictures at an Exhibition" --no-ytdl 'https://r2---sn-25glen7y.googlevideo.com/videoplayback?ipbits=0&gir=yes&lmt=1449557499383412&sparams=clen%2Cdur%2Cei%2Cgir%2Cid%2Cinitcwndbps%2Cip%2Cipbits%2Citag%2Ckeepalive%2Clmt%2Cmime%2Cmm%2Cmn%2Cms%2Cmv%2Cpl%2Crequiressl%2Csource%2Cexpire&mm=31%2C29&ip=86.212.140.188&mn=sn-25glen7y%2Csn-25ge7nsk&ms=au%2Crdu&requiressl=yes&mt=1519150107&pl=17&mv=m&id=o-ANpryLzD7UWNcCDWWQ4Fvl9M_byU1VKiz6gMx5AFTcm2&dur=2013.201&mime=audio%2Fwebm&c=WEB&initcwndbps=557500&itag=251&expire=1519171755&ei=S2SMWp-zGsW_WbndrsgI&key=yt6&keepalive=yes&source=youtube&signature=6592335E75C557E3809495876ABC00E9ABF74230.E2BB779BEE84983F8D1A30144A894C45AFB86A16&clen=34708221&fvip=2&ratebypass=yes'

**Comment by jahlborn** [comment]
so, i reverted back to the release code.  i traced the failure to the metadata call:

DEBUG:root:FOO md failed Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/player.py", line 396, in _launch_player
    metadata = util._get_metadata(song.title)
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/util.py", line 498, in _get_metadata
    metadata = _get_metadata_from_lastfm(t[0], t[1])
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/util.py", line 517, in _get_metadata_from_lastfm
    data = json.loads(resp.read())
  File "/usr/lib/python3.5/json/__init__.py", line 312, in loads
    s.__class__.__name__))
TypeError: the JSON object must be str, not 'bytes'

**Comment by jahlborn** [comment]
If i comment out the line ~392:

    # metadata = util._get_metadata(song.title)
    metadata = None

everything works fine.

**Comment by vn-ki** [comment]
@jahlborn Can you provide the link to the video you are trying to play?

**Comment by jahlborn** [comment]
sure (i was just playing the audio): https://youtu.be/ZAAvPDgKf30

as a side note, something in the code seems to be swallowing the error that gets thrown.  i had to put a try/except directly around that line in order to get the stacktrace into the debug log.  would have been nice if that had been spit out somewhere in at least debug mode.

**Comment by vn-ki** [comment]
Can you try the code my fork https://github.com/vn-ki/mps-youtube.

Switch to the `fix-json-error-in-metadatacall` branch with `git checkout fix-json-error-in-metadatacall`
Run mpsyt with `./mpsyt`.

**Comment by jahlborn** [comment]
Yep, that works.

**Comment by vn-ki** [comment]
@Q3CPMA Can you confirm whether this fix works for you?

**Comment by jahlborn** [comment]
Another helpful change would be to move the "try:" in _get_metadata_from_lastfm() up to encapsulate the whole method.  that way future issues getting metadata won't cause the player to fail.

**Comment by vn-ki** [comment]
> Another helpful change would be to move the "try:" in _get_metadata_from_lastfm() up to encapsulate the whole method. that way future issues getting metadata won't cause the player to fail.

That is a bad coding practice and not a pythonic way of handling things. One would have to use a bare `except` in that case which python actively discourages.

**Comment by jahlborn** [comment]
It's a better practice to have unexpected errors while retrieving metadata (which ultimately isn't necessary for the primary function) to cause the player to not function?

**Comment by atomi** [comment]
Just another confirmation, I had the same issue: `fix-json-error-in-metadatacall` patch fixed it.

Thanks.

**Comment by vn-ki** [comment]
> you could use EnvironmentError or Exception if you don't want a bare except (btw, the code in question is already using a bare except).

They are the same as using bare except. I have fixed the bare except in my PR (though it was not that deadly).

> It's a better practice to have unexpected errors while retrieving metadata (which ultimately isn't necessary for the primary function) to cause the player to not function?

Yes. So that the errors can be fixed. 

But I don't understand what caused the loop issue. It should've raised an exception and stopped execution.

**Comment by jahlborn** [comment]
> Yes. So that the errors can be fixed.

Some possible errors aren't "fixable".  urlopen() can throw IOError if audioscrobbler isn't reachable.  that can fail at any time.

**Comment by atomi** [comment]
I'm still getting weird looping behavior and videos don't play especially when `search_music` is `false` and play through list ie `1-` is used.

**Comment by vn-ki** [comment]
@atomi Please provide the steps to reproduce this. A log would be greatly appreciated.

Is this happening with the `develop` branch or stable release?

**Comment by jahlborn** [comment]
@vn-ki btw, i commented on #785 that the latest version of the patch doesn't fix the problem.

**Comment by atomi** [comment]
@vn-ki 

`
mpsyt version      : 0.2.8
   notes           : released 17 February 2018
pafy version       : 0.5.4 (youtube-dl backend)
youtube-dl version : 2018.02.11
Python version     : 3.5.2 (default, Nov 23 2017, 16:37:01) 
[GCC 5.4.0 20160609]
Processor          : x86_64
Machine type       : x86_64
Architecture       : 64bit, ELF
Platform           : Linux-4.15.1-041501-lowlatency-x86_64-with-LinuxMint-18.1-serena
sys.stdout.enc     : UTF-8
default enc        : utf-8
Config dir         : /home/atomi/.config/mps-youtube
env:TERM           : screen-256color
env:SHELL          : /usr/bin/zsh
env:LANG           : en_US.UTF-8
--
`
I tried with debug but there are no errors it just hangs with intermittent screen refreshes.

**Comment by atomi** [comment]
@vn-ki Can you try doing a `playurl https://www.youtube.com/watch?v=2a4Uxdy9TQY` I think it might be some videos that are hanging mpsyt up.

**Comment by vn-ki** [comment]
> Can you try doing a playurl https://www.youtube.com/watch?v=2a4Uxdy9TQY 

That works perfectly for me. As @jahlborn said, it might be that the patch actually did not fix anything.

**Comment by atomi** [comment]
From latest develop branch

```
Traceback (most recent call last):

  File "/usr/local/bin//mpsyt", line 11, in <module>
    load_entry_point('mps-youtube==0.2.8', 'console_scripts', 'mpsyt')()
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/main.py", line 148, in main
    if matchfunction(i.function, i.regex, userinput):
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/main.py", line 65, in matchfunction
    func(*matches)
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/commands/play.py", line 134, in play_url
    play(override, "1", "_")
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/commands/play.py", line 103, in play
    g.PLAYER_OBJ.play(songlist, shuffle, repeat, override)
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/player.py", line 75, in play
    self._playsong()
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/player.py", line 134, in _playsong
    self._launch_player()
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/player.py", line 290, in _launch_player
    self.send_metadata_mpris()
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/player.py", line 142, in send_metadata_mpris
    metadata = util._get_metadata(self.song.title)
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/util.py", line 501, in _get_metadata
    metadata = _get_metadata_from_lastfm(t[0], t[1])
  File "/usr/local/lib/python3.5/dist-packages/mps_youtube/util.py", line 522, in _get_metadata_from_lastfm
    data = json.loads(resp.decode('utf-8'))
AttributeError: 'HTTPResponse' object has no attribute 'decode'
```

**Comment by Q3CPMA** [comment]
It looks like it works again for me. Youtube-dl may have been the failure's
culprit.

**Comment by vn-ki** [comment]
> From latest develop branch

This should be fixed by d0a3eee.

> It looks like it works again for me. Youtube-dl may have been the failure's
culprit.

Good to know.
